openedx/core/djangoapps/user_authn/views/logout.py

```diff
@@ -83,6 +84,17 @@ def dispatch(self, request, *args, **kwargs):
         delete_logged_in_cookies(response)
 
         mark_user_change_as_expected(None)
+
+        # Redirect to tpa_logout_url if TPA_AUTOMATIC_LOGOUT_ENABLED is set to True and if
+        # tpa_logout_url is configured.
+        #
+        # NOTE: This step skips rendering logout.html, which is used to log the user out from the
+        # different IDAs. To ensure the user is logged out of all the IDAs be sure to redirect
+        # back to <LMS>/logout after logging out of the TPA.
+        if settings.TPA_AUTOMATIC_LOGOUT_ENABLED:
+            if self.tpa_logout_url:
+                return redirect(self.tpa_logout_url)
+
```

Comment:
After the TPA logout endpoint redirects back here, the logout page will still be shown, correct? And I think the page will then offer a button for logging out of SSO: "Click here to delete your single signed on (SSO) session." I'm not sure I'm set up to test this for myself, but I imagine two problems with the code as is:

1. The user will click the button needlessly, since they're already logged out of SSO.
2. If they _do_ click, the button, they'll be redirected back to this logout page, since that's what the SSO is configured to do in order to support this flow.

I wonder if there needs to be an additional parameter on the logout page that prevents that button from appearing. The SSO's post-logout redirect could then instead be `<LMS>/logout?tpa_logout_done=true` or similar.

Reply:
@timmc-edx 
>After the TPA logout endpoint redirects back here, the logout page will still be shown, correct? And I think the page will then offer a button for logging out of SSO: "Click here to delete your single signed on (SSO) session."

Although the logout page will indeed be shown after the TPA logout endpoint redirects back here, the page offering the button for logging out of SSO will not be shown. This is because, the user has [already been logged out](https://github.com/open-craft/edx-platform/blob/13a11388389bfaea6822c5d169edfddcc6ae1b71/openedx/core/djangoapps/user_authn/views/logout.py#L79-L84) from the LMS at this point. The user will just be logged out from the IDAs using the hidden iframes and then redirected to the LMS homepage `/`.

You can check this in the sandbox linked in the testing instructions above.

Reply:
We can also add this flag check in `_show_tpa_logout_link` function to make sure it work properly.
Like:
```suggestion
if bool(target == self.default_target and self.tpa_logout_url) and settings.LEARNER_PORTAL_URL_ROOT in referrer and not settings.TPA_AUTOMATIC_LOGOUT_ENABLED:
     return True
```

Reply:
@mubbsharanwar Thanks. I have now added this check in `_show_tpa_logout_link` as a failsafe.

Reply:
Thanks, makes sense!